cannon/mipsevm/state_mt.go

```diff
@@ -0,0 +1,293 @@
+package mipsevm
+
+import (
+	"encoding/binary"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+// SERIALIZED_THREAD_SIZE is the size of a serialized ThreadState object
+const SERIALIZED_THREAD_SIZE = 166
+
+// THREAD_WITNESS_SIZE is the size of a thread witness encoded in bytes.
+//
+//	It consists of the active thread serialized and concatenated with the
+//	32 byte hash onion of the active thread stack without the active thread
+const THREAD_WITNESS_SIZE = SERIALIZED_THREAD_SIZE + 32
+
+// The empty thread root - keccak256(bytes32(0) ++ bytes32(0))
+var EmptyThreadsRoot common.Hash = common.HexToHash("0xad3228b676f7d3cd4284a5443f17f1962b36e491b30a40b2405849e597ba5fb5")
+
+type ThreadState struct {
+	// Metadata
+	ThreadId uint32 `json:"threadId"`
+	ExitCode uint8  `json:"exit"`
+	Exited   bool   `json:"exited"`
+	// State
+	FutexAddr        uint32     `json:"futexAddr"`
+	FutexVal         uint32     `json:"futexVal"`
+	FutexTimeoutStep uint64     `json:"futexTimeoutStep"`
+	Cpu              CpuScalars `json:"cpu"`
+	Registers        [32]uint32 `json:"registers"`
+}
+
+func (t *ThreadState) serializeThread() []byte {
```

Comment:
nit: worth panicking here with something like `panic("unimplemented")` to avoid accidental use in tests without mocks.

Reply:
good call